### ISSUE TITLE
Support for flow control?

### ISSUE THREAD

**Issue by elliotsayes** [Zekfad/fetch_client#3]
If the listener of fetch's response stream is slower at processing the data than the download speed, then the listener stream's sink grows in size indefinitely.

Is there any way to bridge the nice flow control of dart streams with fetch on the JS side?

**Comment by Zekfad** [comment]
I think it's possible, if we read from ReadableStream only when we actually have listener.
One way of doing that is via generator function with yield, but it needs more investigation.

While I understand your concern, I'm curious about your use case. It's hard to imagine downloading and processing gigabyte of data via browser?

**Comment by elliotsayes** [comment]
Hey 👋

I’m no expert on how dart streams work under the hood, but I think the trick to this is to correctly handle “pause” and “resume” calls from the listening stream. Combined with your of a generator on the JS side probably makes sense here.

As for my use case, it’s end to end encryption of large files from the web app client. Since they are excepted, you can’t download them normally via the browser download, and although decryption is fast thanks to WebCrypto, there is no guarantee that the save medium will be able to keep up with the download speed (e.g. with a slow SD card). I get your point that this is maybe a bit too niche for you to want to worry about :p

**Comment by Zekfad** [comment]
Thanks for your reply, I'll research this and make a fix in a near future.

**Comment by Zekfad** [comment]
I've checked implementation, it seems that is already the case or there are no solution.
Here's implementation:
```dart
  /// Reads stream via [read] and returns chunks as soon as they are available.
  Stream<Uint8List> readAsStream() async* {
    ReadableStreamDefaultReaderChunk chunk;
    do {
      chunk = await read();
      if (chunk.value != null)
        yield chunk.value!;
    } while (!chunk.done);
    return;
  }
```
If you do something like that:
```dart
  final uri = Uri.http('127.0.0.1:8080', '5MB.zip');
  final client = FetchClient(
    mode: RequestMode.cors,
    streamRequests: true,
  );
  final response = await client.send(Request('GET', uri));

  var i = 0;
  await for (final chunk in response.stream) {
    await Future.delayed(Duration(seconds: 1));
    print('chunk: ${++i}');
  }
```
That will print chunks with delays (imagine if that was some long computation).
Buffer indeed will grow beyond what we've read, but that buffer is baked by browser, and fills as soon as data comes from request
There are no way to tell browser to wait before continuing with filling the buffer from request.

Internally we read chucks when they are needed, therefore memory usage wont be more than two times size of file (that is possible only if file is small enough to fit in one chunk, otherwise `read` will spit chunks internally).

**Comment by Zekfad** [comment]
Regarding your use case, I'd suggest you to use WebRTC with WebSockets, because that's the only way I can see peer-to-peer communication.
Actually I cant even see where fetch can be used, as it makes only request to server, not to peer, if you're not using some server as relay proxy, which will make it not an actual peer-to-peer.

**Comment by Zekfad** [comment]
I'll close this for now, if you feel like there are something left, feel free to reopen.

**Comment by elliotsayes** [comment]
> That will print chunks with delays (imagine if that was some long computation). Buffer indeed will grow beyond what we've read, but that buffer is baked by browser, and fills as soon as data comes from request There are no way to tell browser to wait before continuing with filling the buffer from request.

Thanks for looking into this, disappointing to hear but if's baked by the browser as you say then there's nothing that can be done

> Regarding your use case, I'd suggest you to use WebRTC with WebSockets, because that's the only way I can see peer-to-peer communication. Actually I cant even see where fetch can be used, as it makes only request to server, not to peer, if you're not using some server as relay proxy, which will make it not an actual peer-to-peer.

I should have clarified; the use case is data storage (via server) not just peer-to-peer data transfer, so WebRTC is not an option. I would use something like WebSockets but I'm in a weird situation where I don't actually have control over the server and it has to be a regular HTTP request.

Thanks for your help & discussion!

**Comment by Zekfad** [comment]
> it has to be a regular HTTP request

If your server supports [data ranges](https://developer.mozilla.org/en-US/docs/Web/HTTP/Range_requests), you can look into using them for getting chunks of data, rather than full blob in a single request.

**Comment by elliotsayes** [comment]
Great point, that is actually the current approach but it conflicts with the server’s anti-DDoS rate limiting… a weird predicament I know 😅

**Comment by Zekfad** [comment]
You have peculiar conditions I'd say.
You may experiment with chunk sizes to find optimal solution, I'd also suggest to use web workers to cache as much stuff as possible.
Anyway wish you success with your project!

**Comment by elliotsayes** [comment]
Thanks for your help and work on these libs :)

**Comment by Zekfad** [comment]
Just to add some info:
I've recently looked more into JS Streams API, and done some testing, it seems that browser wont crash the app until internal buffer exhaust memory

![image](https://github.com/Zekfad/fetch_client/assets/8970959/594a8dbc-e7ea-4628-8052-c521b4cef4d3)

I've implemented cancelation when you break out of read loop (will publish/push later).
And readers now support backpressure in form of stream pausing (this means, that IF browser send backpressure when you SEND data, it will pause stream, haven't tested with fetch, but works with other API's, so in theory it should at least help to not overwhelm request streaming, where supported).

Also, for your use case, I suggest you to add a processing metric, and server argument (e.g. Header or Query Parameter) for bandwidth limiting, e.g. `limit_rate` in Nginx. This will throttle server upload speed to user, so it could process the data at their own pace. I assume that browser releases parts of the buffer when you done reading and discarding them.
